pkg/cfn/waiter/waiter.go

```diff
@@ -0,0 +1,40 @@
+package waiter
+
+import (
+	"context"
+	"time"
+)
+
+// A Waiter keeps retrying the specified operation until it returns true, or an error.
+type Waiter struct {
+	// NextDelay is the amount of time to wait before the next retry given the number of attempts.
+	NextDelay NextDelay
+
+	// Operation is the function to invoke.
+	Operation func() (bool, error)
+}
+
+// Wait waits for the specified operation to complete.
+func (p *Waiter) Wait(ctx context.Context) error {
+	for attempts := 1; ; attempts++ {
+		done, err := p.wait(ctx, p.NextDelay(attempts))
+		if err != nil {
+			return err
+		}
+		if done {
+			return nil
+		}
+	}
+}
+
+func (p *Waiter) wait(ctx context.Context, d time.Duration) (bool, error) {
+	waitTimer := time.NewTimer(d)
+	select {
+	case <-waitTimer.C:
+		return p.Operation()
+
+	case <-ctx.Done():
+		waitTimer.Stop()
+		return false, ctx.Err()
+	}
+}
```

Comment:
could we not have used the existing waiter from the AWS package? It has the option to configure the timeout via an func https://github.com/aws/aws-sdk-go/blob/60fb75bcde84d6dc06ca60eef07626c542dedd3b/aws/request/waiter.go#L28-L34

Reply:
> It has the option to configure the timeout via an func

It does and our existing code relies heavily on that package, but I found a few issues with it:

1. It uses the concept of acceptors, matchers, PathAllWaiterMatch etc that is too complicated for our requirements. We just need to keep retrying an operation until it's successful. 
2. It's not type-safe. It extracts the value to match against using a JMESPath like `Stacks[].StackStatus`.
3. More importantly, that package was never intended to be used outside of the AWS SDK, as is also indicated by this comment https://github.com/aws/aws-sdk-go/blob/60fb75bcde84d6dc06ca60eef07626c542dedd3b/aws/request/waiter.go#L72-L73

Reply:
For 1 & 2 I haven't seen any problems with us using it, it seems to work well for the majority of our use cases. The comment for 3 is 4 years old and references a func that doesn't exist as far as I can see, so I'm not entirely sure thats true. I don't think we've had any problems using this waiter elsewhere in the codebase so I'm a bit against introducing a third waiter pattern. We already have a 2nd method [here](https://github.com/weaveworks/eksctl/blob/122dc492be096c30ddd4fe7b291838d16132678e/pkg/utils/waiters/waiters.go#L15-L38), could that be refactored to fit this use case if we won't use the AWS waiter? I'm just wary that this introduces a 3rd method for waiting in the codebase, which might be unnecessary

Reply:
>For 1 & 2 I haven't seen any problems with us using it, it seems to work well for the majority of our use cases. 

That's because it has been there forever and is rarely touched, but that doesn't mean it's not technical debt. Anyone new to that code would have to get familiar with acceptors, matchers etc and the fact that it's using JMESPath to extract values. 

>The comment for 3 is 4 years old and references a func that doesn't exist as far as I can see, so I'm not entirely sure thats true.

It's referring to the usage of the type itself and the `WaitUntil*` functions do exist in the service package. 

>We already have a 2nd method here, could that be refactored to fit this use case if we won't use the AWS waiter? I'm just wary that this introduces a 3rd method for waiting in the codebase, which might be unnecessary

I'd argue that the function can be removed in favour of the new one. That one doesn't take a context, accepts the error to return when a timeout error occurs, and there's no need for both a `time.Ticker` and `time.Timer`.

Reply:
> That's because it has been there forever and is rarely touched, but that doesn't mean it's not technical debt. Anyone new to that code would have to get familiar with acceptors, matchers etc and the fact that it's using JMESPath to extract values.

I added some usage for it for addons and it went reasonably smoothly, but I agree its not as intuitive
 
> I'd argue that the function can be removed in favour of the new one. That one doesn't take a context, accepts the error to return when a timeout error occurs, and there's no need for both a `time.Ticker` and `time.Timer`.

Yeah that sounds like a worthwhile refactor